drivers/tee/optee/optee.c

```diff
@@ -0,0 +1,615 @@
+	if (ret) {
+		LOG_ERR("Unable to free shared memory");
+	}
+
+	return (rc) ? rc : ret;
+}
+
+static int optee_cancel(const struct device *dev, uint32_t session_id, uint32_t cancel_id)
+{
+	int rc;
+	struct tee_shm *shm;
+	struct optee_msg_arg *marg;
+
+	rc = tee_add_shm(dev, NULL, OPTEE_MSG_NONCONTIG_PAGE_SIZE,
+			 OPTEE_MSG_GET_ARG_SIZE(0),
+			 TEE_SHM_ALLOC, &shm);
+	if (rc) {
+		LOG_ERR("Unable to get shared memory, rc = %d", rc);
+		return rc;
+	}
+
+	marg = shm->addr;
+	marg->num_params = 0;
+	marg->cmd = OPTEE_MSG_CMD_CANCEL;
+	marg->cancel_id = cancel_id;
+	marg->session = session_id;
+
+	rc = optee_call(dev, marg);
+
+	if (tee_rm_shm(dev, shm)) {
+		LOG_ERR("Unable to free shared memory");
+	}
+
+	return rc;
+}
+
+static int optee_invoke_func(const struct device *dev, struct tee_invoke_func_arg *arg,
+			     unsigned int num_param, struct tee_param *param)
+{
+	int rc, ret;
+	struct tee_shm *shm;
+	struct optee_msg_arg *marg;
+
+	if (!arg) {
+		return -EINVAL;
+	}
+
+	rc = tee_add_shm(dev, NULL, OPTEE_MSG_NONCONTIG_PAGE_SIZE,
+			 OPTEE_MSG_GET_ARG_SIZE(num_param),
+			 TEE_SHM_ALLOC, &shm);
+	if (rc) {
+		LOG_ERR("Unable to get shared memory, rc = %d", rc);
+		return rc;
+	}
+
+	marg = shm->addr;
+	memset(marg, 0, OPTEE_MSG_GET_ARG_SIZE(num_param));
+
+	marg->num_params = num_param;
+	marg->cmd = OPTEE_MSG_CMD_INVOKE_COMMAND;
+	marg->func = arg->func;
+	marg->session = arg->session;
+
+	rc = param_to_msg_param(param, num_param, marg->params);
+	if (rc) {
+		goto out;
+	}
+
+	arg->ret = optee_call(dev, marg);
+	if (arg->ret) {
+		arg->ret_origin = TEEC_ORIGIN_COMMS;
+		goto out;
+	}
+
+	rc = msg_param_to_param(param, num_param, marg->params);
+	if (rc) {
+		arg->ret = TEEC_ERROR_COMMUNICATION;
+		arg->ret_origin = TEEC_ORIGIN_COMMS;
+		goto out;
+	}
+
+	arg->ret = marg->ret;
+	arg->ret_origin = marg->ret_origin;
+out:
+	ret = tee_rm_shm(dev, shm);
+	if (ret) {
+		LOG_ERR("Unable to free shared memory");
+	}
+
+	return (rc) ? rc : ret;
+}
+
+static void *optee_construct_page_list(void *buf, uint32_t len, uint64_t *phys_buf)
+{
+	const uint32_t page_size = OPTEE_MSG_NONCONTIG_PAGE_SIZE;
+	const uint32_t num_pages_in_pl = page_size - 1;
+	uint32_t page_offset = (uintptr_t)buf & (page_size - 1);
+
+	uint8_t *buf_page;
+	uint32_t num_pages;
+	uint32_t list_size;
+
+	/* see description of OPTEE_MSG_ATTR_NONCONTIG */
+	struct {
+		uint64_t pages[OPTEE_NUMBER_OF_ADDR_PER_PAGE - 1];
+		uint64_t next_page;
+	} *pl;
+
+	num_pages = ROUND_UP(page_offset + len, page_size) / page_size;
+	list_size = ceiling_fraction(num_pages, num_pages_in_pl) * page_size;
+
+	pl = k_aligned_alloc(page_size, list_size);
+	if (!pl) {
+		return NULL;
+	}
+
+	memset(pl, 0, list_size);
+
+	buf_page = (uint8_t *)ROUND_DOWN((uintptr_t)buf, page_size);
+
+	for (uint32_t pl_idx = 0; pl_idx < list_size / page_size; pl_idx++) {
+		for (uint32_t page_idx = 0; num_pages && page_idx < num_pages_in_pl; page_idx++) {
+			pl[pl_idx].pages[page_idx] = z_mem_phys_addr(buf_page);
```

Comment:
We have to deal with Xen extended regions here, or at least add a TODO. However, I think it would be better to add some code inside `z_mem_phys_addr` and `Z_MEM_PHYS_ADDR`, specifically changing the checks if we want to work with extended regions

or I can open an extra task for that in our Zephyr repository, because it looks like a generic task

Reply:
I am not sure, why Xen extended regions should affect this. Could you please elaborate?

Reply:
it affects `z_mem_phys_addr` and `Z_MEM_PHYS_ADDR` as far as I understand it is possible to pass to driver memory allocated from xen extended regions and it can happen in other places where used the function and macro mention before:
```
	__ASSERT((addr >= CONFIG_KERNEL_VM_BASE) &&
		 (addr < (CONFIG_KERNEL_VM_BASE +
			  (CONFIG_KERNEL_VM_SIZE))),
		 "address %p not in permanent mappings", virt);
```
even if we have disabled asserts with MMU enabled we can get some bad address for the extended regions
```
#ifdef CONFIG_MMU
#define Z_MEM_VM_OFFSET	((CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_OFFSET) - \
			 (CONFIG_SRAM_BASE_ADDRESS + CONFIG_SRAM_OFFSET))
#else
#define Z_MEM_VM_OFFSET	0
#endif

#define Z_MEM_PHYS_ADDR(virt)	((virt) - Z_MEM_VM_OFFSET)
```

From this driver's perspective, we don't know where the memory is coming from.

Reply:
We definitely don't want asserts to be disabled. Anyways, this is not the problem of this driver. It is more issue of Zephyr memory subsystem. There will be a lot to do during extended regions upstreaming. 

In Linux, for example, Extended regions are implemented as hot-pluggable memory, so Linux mm subsystem can handle them correctly.

Reply:
You wouldn't mind if I open an issue for that in the Zephyr repository, would you?

Reply:
Are you referring to Zephyr Zephyr or Xen-Troops Zephyr?  

If the second one, then of course, create an issue. We need this documented